android/app/src/main/java/app/organicmaps/widget/SpeedLimitView.java

```diff
@@ -0,0 +1,225 @@
+package app.organicmaps.widget;
+
+import android.content.Context;
+import android.content.res.TypedArray;
+import android.graphics.Canvas;
+import android.graphics.Color;
+import android.graphics.Paint;
+import android.graphics.Rect;
+import android.graphics.Typeface;
+import android.util.AttributeSet;
+import android.util.Pair;
+import android.view.View;
+
+import androidx.annotation.ColorInt;
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+
+import app.organicmaps.R;
+import app.organicmaps.util.StringUtils;
+
+public class SpeedLimitView extends View
+{
+  private interface DefaultValues
+  {
+    @ColorInt
+    int BACKGROUND_COLOR = Color.WHITE;
+    @ColorInt
+    int BORDER_COLOR = Color.RED;
+    @ColorInt
+    int ALERT_COLOR = Color.RED;
+    @ColorInt
+    int TEXT_COLOR = Color.BLACK;
+    @ColorInt
+    int TEXT_ALERT_COLOR = Color.WHITE;
+
+    float BORDER_WIDTH_RATIO = 0.1f;
+  }
+
+  @ColorInt
+  private final int mBackgroundColor;
+
+  @ColorInt
+  private final int mBorderColor;
+
+  @ColorInt
+  private final int mAlertColor;
+
+  @ColorInt
+  private final int mTextColor;
+
+  @ColorInt
+  private final int mTextAlertColor;
+
+  @NonNull
+  private final Paint mSignBackgroundPaint;
+  @NonNull
+  private final Paint mSignBorderPaint;
+  @NonNull
+  private final Paint mTextPaint;
+
+  private float mWidth;
+  private float mHeight;
+  private float mBackgroundRadius;
+  private float mBorderRadius;
+  private float mBorderWidth;
+
+  private double mSpeedLimit;
+  @Nullable
+  private String mSpeedLimitStr;
+
+  private double mCurrentSpeed;
+
+  public SpeedLimitView(Context context, @Nullable AttributeSet attrs)
+  {
+    super(context, attrs);
+
+    try (TypedArray data = context.getTheme()
+                                  .obtainStyledAttributes(attrs, R.styleable.SpeedLimitView, 0, 0))
+    {
+      mBackgroundColor = data.getColor(R.styleable.SpeedLimitView_backgroundColor, DefaultValues.BACKGROUND_COLOR);
+      mBorderColor = data.getColor(R.styleable.SpeedLimitView_borderColor, DefaultValues.BORDER_COLOR);
+      mAlertColor = data.getColor(R.styleable.SpeedLimitView_alertColor, DefaultValues.ALERT_COLOR);
+      mTextColor = data.getColor(R.styleable.SpeedLimitView_textColor, DefaultValues.TEXT_COLOR);
+      mTextAlertColor = data.getColor(R.styleable.SpeedLimitView_textAlertColor, DefaultValues.TEXT_ALERT_COLOR);
+      if (isInEditMode())
+      {
+        mSpeedLimit = data.getInt(R.styleable.SpeedLimitView_editModeSpeedLimit, -1);
+        Pair<String, String> speedLimitAndUnits = StringUtils.nativeFormatSpeedAndUnits(mSpeedLimit);
+        mSpeedLimitStr = mSpeedLimit > 0 ? speedLimitAndUnits.first : null;
+        mCurrentSpeed = data.getInt(R.styleable.SpeedLimitView_editModeCurrentSpeed, -1);
+      }
```

Comment:
You're in the `edit mode` - when the view is displayed in the Android Studio XML editor. You shouldn't call any native functions here because there is no application.

Reply:
Hm, why do we code something for Android Studio XML editor ? :)

Reply:
It was my code :)
With these few lines, you'll be able to see how this view works without compiling the app

Reply:
Well, I'd prefer at least comment (or delete) these lines after testing to avoid entropy :)

Reply:
No, don't delete, this behavior is really useful.
And the code itself works as a comment here (`isInEditMode()`).